subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/StrictDependenciesResolveIntegrationTest.groovy

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve
+
+import org.gradle.integtests.fixtures.RequiresFeatureEnabled
+
+@RequiresFeatureEnabled
+class StrictDependenciesResolveIntegrationTest extends AbstractStrictDependenciesIntegrationTest {
```

Comment:
:x: We'll need to add test coverage for the case where the module metadata is inconsistent, like when the `prefer` value is inside the `reject` range. Unit test coverage would be fine. We should probably validate that when parsing.

Reply:
Do you want test when we publish or when we parse? Or both? I'm asking because it implies parsing the selectors, which will add a bit of overhead. I'm also unsure how to do it actually, unless we limit to 2 kind of selectors: "exact" and "range". But then if we do it means that using `latest.release` would be problematic for those who use it today.

Reply:
I'm pretty sure we'll want to split `latest.release` into `+` and `status == 'release'` sometime soon. It makes no sense to encode the status constraint with the version constraint.

We also might want to prevent publishing all of our crappy range syntaxes to the new metadata, and translate to some common, unambiguous notation. e.g. we could prohibit `1+`, and we might want to come up with a better solution for `1.+` (or is it strictly equivalent to `[1.0,)`?).

I'm OK with only validating `prefer/reject` on publish for now, even though I think eventually we'll want to have a richer, more strongly typed model, which will involve more validation on parsing.

Reply:
So you would only validate `ExactVersionSelector` and `RangeVersionSelector`? I mean, if we have:

- accept `i`, reject `[x,y]`, we can extract `i` from `ExactSelector` and check that the range doesn't match it
- accept `[i,j]` and reject `[x,y]`, we can check that the ranges do not intersect

But for other selectors, it doesn't seem possible. For example:

- accept `latest` (not `latest.release`)
- accept `1+`
- accept `1.+`

Since all of them do not provide "a version" or something that can be checked against the other selector.

Reply:
Let's ignore the 'status' for now, since that shouldn't really be part of the version matcher at all.
Then, I think we can translate each version string input into either:
a) An exact version
b) A well-defined version range.
c) A deprecated, opaque version that we can't fully translate

For a & b, we can calculate the exact complement to use for `reject`.

input | prefer | reject | notes
------ | ------  | ------- | ------
1.2 | 1.2 | (,1.2[ 
strictly 1.2 | 1.2 | (,1.2[ + ]1.2,)
[1.2,) | [1.2,) | (,1.2[
strictly [1.2,) | [1.2,) | (,1.2[ | strictly has no meaning
(1,2,) | ]1,2,) | (,1.2] 
strictly (1.2,) | ]1.2,) | (,1.2] | strictly has no meaning
[2.0,3.0) | [2.0,3.0[ | (,2.0)
strictly [2.0,3.0) | [2.0,3.0[ | (,2.0) + [3.0,) 
1.+ | [1.0, 2) | (,1.0[ + [2,) | Not certain this is exactly the same meaning
1+ | n/a | n/a | No translation possible: cannot publish to module metadata
| + | (,) | (,) | None preferred (will use latest), but also none rejected
latest | (,) | (,) | Exactly the same as '+'

Reply:
@melix Let's think about this table above and possibly add it as a proposal for the module metadata spec. This is as good a time as any to get rid of crappy old version range behaviours and syntaxes.

Reply:
Right, it seems there's an overlap with #3454, let's fix there.